### PR TITLE
install py or src not both; no silent installdir fallback

### PR DESCRIPTION
**Pull request by sbailey**
Our two hybrid python/C++ packages (fiberassign and specex) have now converted to using `python setup.py install` for installations, including compiling the C++ code.  This was creating a fake-failure problem for desiInstall which was noticing the `src/` directory and trying to run a non-existent Makefile.  This PR updates the installation code to use build type "py" OR "make" OR "src" but not "py" AND "src".  I left the "make" and "src" logic in place even though we don't have any packages using them now, but if the `setup.py` file is found it uses that and then moves on.

This also fixes a bug where if the `desiInstall --root ...`directory didn't exist at NERSC, the code was silently falling back to the default NERSC install location instead of warning the user that their requested installation location doesn't exist and stop.

With current master:
```
[cori12 temp] desiInstall -r $SCRATCH/desi/temp fiberassign 2.1.1
CRITICAL:install.py:866:install:2021-02-11T15:15:02: Error during compile: make: *** src: No such file or directory.  Stop.
```
That produces a useable installation, albeit without the permissions fully corrected.

With this branch there are no errors, fully successful install.

### CODE REVIEW

**Comment by weaverba137** [comment]
I think this is fine at first glance, but there is also the desiInstall.rst documentation that will need to be changed.  Could you do that?

**Comment by weaverba137** [comment]
Even though these are vestigial, is the remaining logic "make" XOR "src"?  That is, all possible build types are mutually exclusive?

**Comment by sbailey** [comment]
My intension was to make all build types exclusive (make and src already were, but previously they could be combined with py).  I'll update docs.

**Comment by weaverba137** [comment]
Good, thank you.  I'll do a formal review soon, but it might have to wait until tomorrow morning.

File: doc/desiInstall.rst [hunk 1]
```diff
@@ -224,6 +222,9 @@ src
     :command:`desiInstall` will attempt to execute :command:`make -C src all`.  This build type
     *is* mutually exclusive with 'make', but is not mutually exclusive with
```
Comment (weaverba137) [inline doc/desiInstall.rst:223]: This contradicts the statement above that the build types are mutually exclusive.

File: doc/desiInstall.rst [hunk 2]
```diff
@@ -224,6 +222,9 @@ src
...
     the other types.
+plain
+    If no other build type is detected, the downloaded code
+    is simply copied to the final install directory.
 
 **It is the responsibility of the code developer to ensure that these
```
Comment (weaverba137) [inline doc/desiInstall.rst:229]: Since the build types are now mutually exclusive, perhaps we should rephrase: "It is the responsibility of the code developer to understand these build types and choose the one appropriate for the package being developed."

File: py/desiutil/install.py [hunk 3]
```diff
@@ -573,17 +573,16 @@ def set_install_dir(self):
...
-        except KeyError:
-            self.nersc = None
-        if self.options.root is None or not os.path.isdir(self.options.root):
-            if self.nersc is not None:
-                self.options.root = self.default_nersc_dir()
-            else:
-                message = "Root install directory is missing or not set."
-                self.log.critical(message)
-                raise DesiInstallException(message)
+        self.nersc = os.getenv('NERSC_HOST') #- will be None if not set
```
Comment (weaverba137) [inline py/desiutil/install.py:576]: Is this particular change really necessary?  I really hate `os.getenv()` because it never throws an exception, so it's much harder to tell when a variable is undefined.

File: py/desiutil/install.py [hunk 4]
```diff
@@ -573,17 +573,16 @@ def set_install_dir(self):
...
+        if self.options.root is None and self.nersc is not None:
+            self.options.root = self.default_nersc_dir()
+
+        if (self.options.root is None or not os.path.isdir(self.options.root)) \
```
Comment (weaverba137) [inline py/desiutil/install.py:580]: Use parenthesis to break across lines.  The backslash is ancient history.

File: py/desiutil/test/test_install.py [hunk 5]
```diff
@@ -326,7 +326,7 @@ def test_set_install_dir(self):
         with patch.dict('os.environ', {'NERSC_HOST': 'edison', 'DESI_PRODUCT_ROOT': 'FAKE'}):
             del environ['DESI_PRODUCT_ROOT']
             test_code_version = 'test-blat-foo'
-            options = self.desiInstall.get_options(['desiutil', test_code_version])
+            options = self.desiInstall.get_options(['desiutil', test_code_version, '--test'])
```
Comment (weaverba137) [inline py/desiutil/test/test_install.py:329]: Could you explain in more detail why this change was necessary?

**Review by weaverba137 — APPROVED** [review]
I found a few minor issues to discuss.

**Review by weaverba137 — CHANGES_REQUESTED** [review]
Please see my other review.  This is just to ensure that a "Request changes" result is returned.

File: doc/desiInstall.rst (hunk 1 above)
Comment (sbailey) [reply doc/desiInstall.rst]: Thanks for catching that; I corrected it.

File: doc/desiInstall.rst (hunk 2 above)
Comment (sbailey) [reply doc/desiInstall.rst]: Good; rephrased with your suggestion.

File: py/desiutil/install.py (hunk 3 above)
Comment (sbailey) [reply py/desiutil/install.py]: OK.  I reverted the `os.getenv()` to use `try/except` around `os.environ` again, while noting that simply catching the exception and setting the variable to `None` is exactly the same thing that `os.getenv` does if the environment variable isn't defined.  I agree that `os.getenv()` is a landmine if the code assumes that the variable is defined and would choke if silently given a `None`, but in this case `None` is the correct fallback, so `os.getenv` seemed simpler.  (And now I've spent 7 lines of text explaining why I prefered 1 line of code over 4... :)

I think you were only commenting on the `os.getenv()` line, but just to be clear: the rest of the changes here are important, because they are fixing the problem where desiInstall was silently falling back to the default NERSC location if the user-specified root directory didn't exist.

File: py/desiutil/install.py (hunk 4 above)
Comment (sbailey) [reply py/desiutil/install.py]: I have a gray beard and have been using python since before numpy existed.   But ok, parentheses it is...

File: py/desiutil/test/test_install.py (hunk 5 above)
Comment (sbailey) [reply py/desiutil/test/test_install.py:329]: `set_install_dir()` will now raise an exception if the installation directory doesn't exist, which was tripping up the test that was using the fallback to the NERSC default location, which doesn't exist in the test environment.  When actually installing at NERSC, I think it should be an error if the default location doesn't exist, so I used the `--test` hooks to prevent the existence check exception in `set_install_dir` while testing.

**Comment by sbailey** [comment]
@weaverba137 thanks for the comments / suggestions.  I think I have addressed all of them.  Please re-review.

File: py/desiutil/install.py (hunk 3 above)
Comment (weaverba137) [reply py/desiutil/install.py]: I realize both old and new have the same result. So why change it?

Cosmetic changes like this distract from the real changes. I've heard that before in some of my own pull requests.

Although there apparently is no active unit test on this bit of code, the try: clause would potentially be something that is deliberately triggered in a unit test.

File: py/desiutil/test/test_install.py (hunk 5 above)
Comment (weaverba137) [reply py/desiutil/test/test_install.py:329]: OK.  I'll take a look and see if the unit test can be made a bit more robust.

File: py/desiutil/install.py (hunk 3 above)
Comment (sbailey) [reply py/desiutil/install.py]: Noted and agreed about mixing cosmetic changes distracting from the substantive changes.  In this case it was the result of several iterations of sorting out what this code should do and various attempts to find the best order of if/then try/except for what should be checked in what order, and it landed back there in the N-1 iteration.  I have reverted it back your preferred method.

**Review by weaverba137 — APPROVED** [review]

**Comment by weaverba137** [comment]
OK to merge now.  I need to do an overhaul of the desiutil testing infrastructure anyway but I will punt that to a separate issue.  I can create a tag after merge if you're ready for that.

**Comment by sbailey** [comment]
Thanks, merging now.  I may have another PR or two for desiutil today prior to a tag for 21.2 / Cascades, so no tag yet.
